eureka/S3_data_reduction/niriss.py

```diff
@@ -459,12 +473,119 @@ def wave_NIRISS(wavefile, meta):
     return meta
 
 
-def fit_bg(data, meta):
+def fit_bg(data, meta, n_iters=3, readnoise=5, sigclip=[4,2,3], isplots=0):
     """
     Subtracts background from non-spectral regions.
 
-    # want to create some background mask to pass in to 
-      background.fitbg2
+    Parameters
+    ----------
+    data : object
+    meta : object
+    n_iters : int, optional
+       The number of iterations to go over and remove cosmic
+       rays. Default is 3.
+    readnoise : float, optional
+       An estimation of the readnoise of the detector.
+       Default is 5.
+    sigclip : list, array, optional
+       A list or array of len(n_iiters) corresponding to the
+       sigma-level which should be clipped in the cosmic
+       ray removal routine. Default is [4,2,3].
+    isplots : int, optional
+       The level of output plots to display. Default is 0 
+       (no plots).
+
+    Returns
+    -------
+    data : object
+    """
+    def dirty_mask(img, boxsize1=70, boxsize2=60):
+        """Really dirty box mask for background purposes."""
+        order1 = np.zeros((boxsize1, len(img[0])))
+        order2 = np.zeros((boxsize2, len(img[0])))
+        mask = np.ones(img.shape)
+        
+        for i in range(img.shape[1]):
+            s,e = int(meta.tab2['order_1'][i]-boxsize1/2), int(meta.tab2['order_1'][i]+boxsize1/2)
+            order1[:,i] = img[s:e,i]
+            mask[s:e,i] = 0
+
+            s,e = int(meta.tab2['order_2'][i]-boxsize2/2), int(meta.tab2['order_2'][i]+boxsize2/2)
+            try:
+                order2[:,i] = img[s:e,i]
+                mask[s:e,i] = 0
+            except:
+                pass
+
+        return mask
+
+    box_mask = dirty_mask(data.median)
+    data = fitbg3(data, np.array(box_mask-1, dtype=bool), 
+                  n_iters, readnoise, sigclip, isplots)
+    return data
+
+
+def fit_orders(data, meta):
     """
+    Creates a 2D image optimized to fit the data. Currently
```

Comment:
Can you help me understand what this function is doing?

Reply:
So basically what it's doing is similar to what I had slack'd you about (e.g. profile fitting for optimal extraction). So what you get out of this function are 2 images: (1) has a weighted mask for the first order and (2) has a weighted mask for the second order. Each of the profiles are now are being approximated with a Gaussian of varying amplitudes.

I hope this image will help explain what the final output of the code is.
Left top: example NIRISS exposure
Left bottom: the 2 fitted masks co-added
Right column: the differences in profile shapes at different rows, where the orange is the data profile, and the black is the profile from the fitted mask 

![niriss_2d_mask](https://user-images.githubusercontent.com/5041100/153113887-9de86b03-7f1d-4b00-834f-1576b9e5758d.png)

The reason why it takes so long is because it's fitting the amplitudes and the standard deviations of the gaussian models to the data (I also want it to be flexible enough to fit the center value, but I need to make it faster before adding more variables. Also, this way the order tracing done before is treated as an initial guess, rather than ground truth. But I digress....)

Reply:
Also, here's a comparison between spectra extracted (after all processing) with a box (blue) vs. with the weighted mask derived from this function (black)

![order_comparison](https://user-images.githubusercontent.com/5041100/153115585-7d76060c-8c60-4374-a338-b315efa9b489.png)

Reply:
Thanks for the extra info.  I'll take a second look at the code this morning to piece it all together.
You should check out `fitgaussians()` in [gaussian.py](https://github.com/kevin218/Eureka/blob/main/eureka/lib/gaussian.py).  It may prove to be faster than you current implementation.  
When it comes to fitting the center, it's often good to use the previous best fit center as your starting point for the next columns; however, using the trace would probably be just as good, so I leave that up to you.

Reply:
Alternatively, if you want to focus your time on other routines that yield more realistic profiles, we can simply leave this one as is for now.

Reply:
Ah, I see! I think there are definitely good lessons in `fitgaussian()` to be implemented here and try to wrap them into my implementation in the next PR, with a better profile than a gaussian

Reply:
So I managed to throw together a routine with `scipy.optimize.leastsqr` that's certainly faster and mildly different from the function in `gaussian.py`. The results between that best-fit model and mine seems, minimal (see below)

The lefthand column has 2 masks: the top is with my optimization, the bottom is with `scipy.optimize`. The top mask extracted the blue spectra and the bottom mask extracted the black spectra (to the right)

![order_comparison](https://user-images.githubusercontent.com/5041100/153529967-759c10d0-2247-43b3-a485-62c799ed69cb.png)

What do you think?

Reply:
Is there any way to know the truth? The significant difference in flux between the two methods is worrisome.

Reply:
I've asked Joe for the ground truth of what was injected in the simulation. When he gets those to me, that will probably instill some confidence in one method over the other

Reply:
I'm worried this `so.leastsqr` method is unreliable, since I reran everything this morning and am now getting different results.... I think the current version I have pushed is better for now (even if it's slower)

Reply:
I've pushed a version now with both the old method (`fit_orders`) and a faster version with scipy.optimize (`fit_orders_fast`). I think this is the best I can do right now before I get the ground truth results to compare to.

Reply:
Agreed!